lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -168,36 +182,158 @@ data BalanceInsufficientError = BalanceInsufficientError
+        out { Tx.tokens = augmentBundle (Tx.tokens out) }
+      where
+        augmentBundle bundle =
+            if TokenBundle.getCoin bundle == Coin 0 then
+                bundle { coin = minCoinValueFor (view #tokens bundle) }
+            else
+                bundle
+
+    mkInputsSelected :: UTxOIndex -> NonEmpty (TxIn, TxOut)
+    mkInputsSelected =
+        fromMaybe invariantSelectAnyInputs . NE.nonEmpty . UTxOIndex.toList
+
     balanceAvailable :: TokenBundle
     balanceAvailable = view #balance utxoAvailable
 
     balanceRequired :: TokenBundle
-    balanceRequired = F.foldMap (view #tokens) outputsToCover
-
-    mkResult :: SelectionState -> SelectionResult
-    mkResult SelectionState {selected, leftover} =
-        case NE.nonEmpty (UTxOIndex.toList selected) of
-            Nothing ->
-                unableToSelectAnyInputsError
-            Just inputsSelected ->
-                SelectionResult
+    balanceRequired = F.foldMap (view #tokens) outputsConsidered
+
+    -- There's a chicken-and-egg situation when it comes to calculating
+    -- transaction fees. On the one hand, we need to know the shape of the final
+    -- transaction to calculate its cost. But in order to construct the
+    -- transaction, we need to know what its cost is.
+    --
+    -- So, in order to not duplicate the logic from 'makeChange', we first
+    -- calculate a pre-selection considering the case where we have no fees to
+    -- pay, and no minimum value. This is guaranteed to succeed and will yield
+    -- the final shape of the change outputs (modulo amounts).
+    --
+    -- From there, we can successively re-iterate and construct a final
+    -- unbalanced selection that is leaving enough money out for fees. For this
+    -- method to work (efficiently), the function estimate the cost of a
+    -- selection MUST NOT depends on change token quantity. That is, adding more
+    -- change to an output should not make it more expensive.
+    mkPreSelection
+        :: NonEmpty (TxIn, TxOut)
+        -> SelectionResult [AssetId]
+    mkPreSelection inputsSelected =
+        let
+            changeGenerated = maybe
+                (invariantResultWithNoCost inputsSelected)
+                transformChange
+                (makeChange noMinimumCoin noCost
+                    extraCoinSource
+                    (view #tokens . snd <$> inputsSelected)
+                    (view #tokens <$> outputsConsidered)
+                )
+        in
+            SelectionResult
+                { inputsSelected
+                , outputsConsidered
+                , changeGenerated
+                }
+      where
+        noMinimumCoin :: TokenMap -> Coin
+        noMinimumCoin = const (Coin 0)
+
+        noCost :: Coin
+        noCost = Coin 0
+
+        transformChange :: NonEmpty TokenBundle -> NonEmpty [AssetId]
+        transformChange =
+            fmap (fmap fst . snd . TokenBundle.toFlatList)
```

Comment:
**Proposal:** Narrow the range of this function from `SelectionResult [AssetId]` to `NonEmpty (Set AssetId)`, and give it a name that indicates it only generates initial change assets.

**Justification**: This function doesn't generate new values for `inputsSelected` or `outputsConsidered`, and the values of these fields in the `SelectionResult` can only ever be what was supplied by the caller. But this isn't made apparent by the type signature, which (arguably) increases the work required to understand this function.

By narrowing the function's range, we can make it clear to readers that this function is only responsible for generating an initial value for `changeGenerated`, and even simplify the definition:

```suggestion
    mkInitialChangeAssets :: NonEmpty (TxIn, TxOut) -> NonEmpty (Set AssetId)
    mkInitialChangeAssets inputsSelected = maybe
        (invariantResultWithNoCost inputsSelected)
        (fmap TokenBundle.getAssets)
        (makeChange noMinimumCoin noCost
            extraCoinSource
            (view #tokens . snd <$> inputsSelected)
            (view #tokens <$> outputsCovered)
      where
        noMinimumCoin :: TokenMap -> Coin
        noMinimumCoin = const (Coin 0)

        noCost :: Coin
        noCost = Coin 0
```

The call site then becomes:
```patch
+        let inputsSelected = mkInputsSelected selected
+        let changeGenerated = mkInitialChangeAssets inputsSelected
         let preSelection :: SelectionResult (Set AssetId)
-            preSelection = mkPreSelection (mkInputsSelected selected)
+            preSelection = SelectionResult
+                {inputsSelected, outputsCovered, changeGenerated}
```

Reply:
Yes, I am not yet satisfied with this type signature and the many transformations happening here. I left a small note in the code about it and I intend to revise it. Re-using the `SelectionResult` type may not be the right path to follow at all.